University2.scala

```diff
@@ -0,0 +1,244 @@
+package org.apps.gist
+package university
+
+import scalaz._, Scalaz._
+
+/**
+ * State algebras
+ */
+
+trait Getter[P[_],S]{
+  def apply(): P[S]
+}
+
+object Getter {
+  def apply[P[_], S](get: P[S]): Getter[P, S] =
+    new Getter[P, S] { def apply() = get }
+}
+
+trait Setter[P[_],S]{
+  def apply(t: S): P[Unit]
+}
+
+object Setter {
+  def apply[P[_], S](put: S => P[Unit]): Setter[P, S] =
+    new Setter[P, S] { def apply(s: S) = put(s) }
+}
+
+trait Field[P[_],S]{
+  val get: Getter[P,S]
+  val put: Setter[P,S]
+
+  def modify(f: S => S)(implicit B: Bind[P]): P[Unit] = 
+    get() >>= { s => put(f(s)) } 
+}
+
+object Field {
+  def apply[P[_], S](getter: Getter[P, S], setter: Setter[P, S]) =
+    new Field[P, S] { val get = getter; val put = setter }
+}
+
+trait ListP[Alg[_[_],_],P[_],S]{
+  val apply: P[List[Alg[P,S]]]
+
+  def traverse[T](
+      f: Alg[P,S] => P[T])(implicit 
+      M: Monad[P]): P[List[T]] = 
+    apply >>= { _ traverse[P,T] f }
+
+  def traverseZip[A,T](
+      values: List[A])(
+      f: (Alg[P,S],A) => P[T])(implicit 
+      M: Monad[P]): P[List[T]] =
+    apply >>= { algs => 
+      (algs zip values) traverse f.tupled
+    }
+}
+
+object ListP {
+  def apply[Alg[_[_], _], P[_], S](ap: P[List[Alg[P, S]]]) =
+    new ListP[Alg, P, S] { val apply = ap }
+}
+
+trait StdTraverse[P[_],S] extends ListP[Field,P,S]{
+  def getAll(implicit M: Monad[P]): P[List[S]] =
+    traverse(_.get())
+
+  def putAll(values: List[S])(implicit M: Monad[P]): P[List[Unit]] =
+    traverseZip(values)(_ put _)
+}
+
+
+/** 
+ * Utilities & Combinators 
+ */
+
+object Util {
+
+  type StateField[S, A] = Field[State[S, ?], A]
+
+  object StateField {
+
+    def apply[S, A](g: S => A, p: A => S => S): StateField[S, A] =
+      new Field[State[S, ?], A] {
+        val get = Getter(State.gets(g))
+        val put = Setter(a => State(s => (p(a)(s), ())))
+      }
+
+    def id[S]: StateField[S, S] = Lens.id[S]
+  }
+
+  type Lens[S, A] = State[A, ?] ~> State[S, ?]
+
+  object Lens {
+    def id[S]: Lens[S, S] = NaturalTransformation.refl
+  }
+
+  implicit def lens2StateField[S, A](ln: Lens[S, A]): StateField[S, A] =
+    new Field[State[S, ?], A] {
+      val get = Getter(ln(State.get)) 
+      val put = Setter(a => ln(State.put(a)))
+    }
+
+  implicit def stateField2Lens[S, A](fl: StateField[S, A]): Lens[S, A] =
+    λ[State[A, ?] ~> State[S, ?]] { sta =>
+      State { s =>
+        val (a, x) = sta.run(fl.get().eval(s))
+        (fl.put(a).exec(s), x)
+      }
+    }
+  
+  // Maps the interpretation of an algebra. 
+  trait AlgFunctor[Alg[_[_], _]] {
```

Comment:
This class is nice!

Just a quick comment: we should state clearly which are the differences between `AlgFunctor` and `HFunctor`, since the signature is the same. Different laws?

Reply:
https://hackage.haskell.org/package/category-extras-0.52.1/docs/Control-Functor-HigherOrder.html

Reply:
Ouch! I was pretty sure that `HFunctor` didn't have the additional parameter in the algebra, but clearly I was wrong. Thereby, I'm afraid I have reinvented `HFunctor`. I'll try to reimplement it using this abstraction (btw, is it available in Scalaz? If so, the name seems to be different!).

Reply:
Are StateAlgebras HFunctors? The resulting algebra has to be a functor, so it should map the state. Do we have that?

In scalaz we have MonadTrans, which is something like HMonad, I guess. The method hoist is implemented, not primitive.